src/main/resources/lib/translate/queue.ts

```diff
@@ -1,41 +1,107 @@
 import * as taskLib from '/lib/xp/task';
 import cron from '/lib/cron';
 
-import {TRANSLATION_POOL_SIZE} from '../config';
-import {logDebug, LogDebugGroups} from '../logger';
 
-type QueuedTask = {
+
+import { TRANSLATION_POOL_SIZE } from '../config';
+import { logDebug, LogDebugGroups } from '../logger';
+
+
+
+
+
+type Task = {
     description?: string;
     func: () => void;
     onError: () => void;
 };
 
+type Group = {
+    id: string;
+    tasks: Task[];
+};
+
 class TaskQueue {
+    private groups: Group[] = [];
+
+    addTask(task: Task, groupId: string): void {
+        for (const group of this.groups) {
+            if (group.id === groupId) {
+                group.tasks.push(task);
+                return;
+            }
+        }
+
+        this.groups.push({
+            id: groupId,
+            tasks: [task],
+        } satisfies Group);
+    }
+
+    takeTask(): Task | undefined {
+        if (this.groups.length === 0) {
+            return undefined;
+        }
+
+        const nextGroupToExecute = this.groups.shift();
+        if (!nextGroupToExecute) {
+            return undefined;
+        }
+
+        const nextTask = nextGroupToExecute.tasks.shift();
+        if (!nextTask) {
+            return undefined;
+        }
+
+        if (nextGroupToExecute.tasks.length > 0) {
+            // adding the group to the back of the queue if it still has tasks
+            this.groups.push(nextGroupToExecute);
+        }
+
+        return nextTask;
```

Comment:
Overall, looks good! I'm just thinking we could move it to a separate file, like `TaskQueue.ts`, either under the same `/translate` folder or even under `/utils`. There's a lot of code in the queue right now that we don't really care about.

If you go with this approach, it might also be a good idea to drop the `*ToExecute` from the variable name, since it doesn't really relate to execution in this class.